### PR TITLE
better support for rtl languages

### LINKED ISSUE #89: Handling RTL
**Description**
For now the UI is not yet supporting right-to-left display mechanism of the UI such as the login page and the top menu.

**Expected behavior**
Depending on the language set in the the user's browser `user-agent` the UI should be displayed the right way such as in arabic, hebrew an farsi languages.

Best regards,

### PR DESCRIPTION
**Pull request by DerMolly**
fixes #89 
 @BoFFire and  @ahangarha should check it out and give feedback if possible…

image

image

### CODE REVIEW

**Comment by ahangarha** [comment]
it looks good. but look at this screenshot:

image

I think in the menu, icons should appear before text, so in LTR it should go to left and in RTL it should go to right. That can happen by adding `dir="auto"` to those elements.

**Comment by DerMolly** [comment]
We can't really use `dir=auto` unfortunately, because that only works with the browsers language settings. The App should be able to change via the language picker. Then I just set the whole page to `dir=rtl` if the chosen language is arabic (I think that's our only RTL language right now)...
I could make it that those entries are always ltr, but maybe we should just add an arabic translation for theses texts...

File: src/initializers/i18n.ts [hunk 1]
```diff
@@ -49,4 +49,27 @@ export const setUpI18n = async (): Promise<void> => {
     })
 
   moment.locale(i18n.language)
+  setHTMLDirection(shortenLanguageCode(i18n.language))
+}
+
+export const setHTMLDirection = (language: string): void => {
+  console.log(language)
```
Comment (ErikMichelson) [inline src/initializers/i18n.ts:56]: Why do we output/log this?

File: src/components/landing/layout/footer/language-picker.tsx [hunk 2]
```diff
@@ -2,22 +2,17 @@ import moment from 'moment'
...
-
-  const shortenLanguageCode = (language: string): string => {
-    const languageWithoutArea = language.substr(0, 2)
-    if (languageWithoutArea === 'zh') {
-      return language
-    } else {
-      return languageWithoutArea
-    }
+    const language = event.currentTarget.value
+    console.log('onChangeLang', language)
```
Comment (ErikMichelson) [inline src/components/landing/layout/footer/language-picker.tsx:12]: Same question as below: Why do we output/log this? This isn't production relevant and bloating the code unnecessary.

File: src/initializers/i18n.ts (hunk 1 above)
Comment (DerMolly) [reply src/initializers/i18n.ts:56]: that's 'debug' code

File: src/components/landing/layout/footer/language-picker.tsx (hunk 2 above)
Comment (DerMolly) [reply src/components/landing/layout/footer/language-picker.tsx:12]: that's 'debug' code

File: src/initializers/i18n.ts (hunk 1 above)
Comment (ErikMichelson) [reply src/initializers/i18n.ts:56]: Yes, and debug-code should not be committed, should it?

Comment (mrdrogdrog) [reply src/initializers/i18n.ts]: dont use console.log. Use console.debug please.

**Comment by ahangarha** [comment]
@DerMolly 

See. As a person whose primary languge is Persian which is RTL, there are many cases in which I have LTR texts. Beside this, There are some cases in which, here are English texts in the app which are not translated or not supposed to be translated.

I don't know what is the issue for not adding `dir="auto"` but If you want, I can send a PR. Just I don't know from where the dropdown menu gets `text-align: left;`. I can manage it by adding `text-align: start;` for the relevant elements in styles.

Let me know what to do.

**Comment by DerMolly** [comment]
I think I misunderstood what exactly `dir='auto'` does. I added it to a toplevel div and some elements which have intentionally english text or are just not yet translated. Best case scenario we would add  `dir='auto'` to each element via css, but that's unfortunately not possible. At least that's my understanding of https://github.com/w3c/csswg-drafts/issues/2091. 

We should add `dir='auto'` to more elements, to ensure a better translation progress for our next RTL language

**Comment by mrdrogdrog** [comment]
You don't have to add it to every element. The "dir" attribute will be inherited!

**Comment by DerMolly** [comment]
Since `dir='auto'` performes a language direction detection and we could have untranslated texts, we need to add `dir='auto'` not to every element (that would have just been the easiest way if it could be done via CSS), but to each element where a direction change may occur. For arabic that is currently done and works fine (I think), but new RTL languages in the future, which are not 100% translated yet, would need to have extra `dir='auto'` thrown into the mix…

**Comment by mrdrogdrog** [comment]
No, you don't have to put auto to a lower element, if a dir=auto has it already. All we have to do is to set dir=auto to the root div

**Comment by mrdrogdrog** [comment]
image

**Comment by mrdrogdrog** [comment]
Or is there any problem with this representation?

**Comment by DerMolly** [comment]
There is in the menu as @ahangarha pointed out the first to paragraphs should have the icon on the left, as it currently is in this branch and because the text direction changes from rtl to ltr in this case you either need to add `dir='ltr'` (as done for the CodiMD and logo on the intro page) or (because we don't know if the text is truly translated) add an extra `dir='auto'` to those entries to possibly change the text direction.

**Comment by mrdrogdrog** [comment]
So you want mixed directions?

**Comment by DerMolly** [comment]
I think it needs to be possible

**Comment by mrdrogdrog** [comment]
Well... That could be difficult, but not impossible. 
Not every component we use (especially from react-bootstrap) has a direction prop. But we can wrap the content of these elements in e.g. spans and add a direction to them.

**Comment by DerMolly** [comment]
The question is do we want to do that now or do we want to work the scout approach?

**Comment by mrdrogdrog** [comment]
IMHO : I would wait. Currently we are altering the UI very often. It's easier to do such extra tasks later. But I would still include it into the 2.0.

**Comment by mrdrogdrog** [comment]
First I would work out the complete 2.0 UI and then optimize it for RTL.

**Comment by DerMolly** [comment]
Then this should be merged and an extra issue should be opened to fix this when the code is final or at least close to that. In the mean time new code should include `dir='auto'`. I'll merge that later, in case there are other voices to heard here and open the issue.

**Comment by mrdrogdrog** [comment]
Can't we just leave this issue open and add `dir=auto` to the HTML tag for now?

**Comment by mrdrogdrog** [comment]
Please rebase the pr first

**Comment by ahangarha** [comment]
Guys,

I explained above that unless you are sure about whole website contains only LTR or RTL text, it is mistake to think applying direction in top level (root) is going to give desirable result.

Virtually, all the text containing elements should have `dir="auto"` so based on its content, it would show the text in proper direction. This is important to have proper bidi support.

If CSS has `direction:auto;` it would be so easy to implement. But for now, using `dir="auto"` is the only way. The other workaround is to do something similar to what my firefox addon does: run JS on page and add `dir="auto"` and `text-align: start;` to all p, a, hx, ul, ol, li, pre.

**Comment by mrdrogdrog** [comment]
That's why I propose, that we first finish the UI and then, as one of the last tasks for 2.0, add optimizations for e.g. rtl.

**Comment by ahangarha** [comment]
Then keep `dir="auto"` in top level. I try to gradually come and contribute to bidi support.

**Comment by SISheogorath** [comment]
Since react is a component-based structure, shouldn't each component become RTL-aware on its own? In this case PRs could happen for each component separately and therefore being less likely to cause problems with changes.

**Pull request merged**